### PR TITLE
Active learning

### PR DESCRIPTION
**Pull request by vitartas**
Active learning modifications:
1. Possibility to use PlumedBias during active learning
2. Possibility to add ASE constraints during active learning
3. Option to specify at which iteration to start applying the bias
4. If using metadynamics PLUMED bias, an option to inherit metadynamics bias from previous active learning iterations

Other modifications:
1. Possibility to specify a PlumedBias to Metadynamics
2. New and adjusted properties to ConfigurationSet
3. Possibility to use ASE constraints in run_mlp_md()

### CODE REVIEW

**Comment by t-young31** [comment]
hey @vitartas – hope you don't mind but I'm going to review this. hopefully some useful suggestions and likely lots of ignorable rubbish!

File: mlptrain/configurations/configuration.py [hunk 1]
```diff
@@ -39,6 +39,8 @@ def __init__(self,
         self.energy = Energy()
         self.forces = Forces()
 
+        self.plumed_coordinates = None
```
Comment (t-young31) [inline mlptrain/configurations/configuration.py:42]: this needs a type hint for what it is, and probably (as much as I don't like them) an inline comment
```suggestion
        self.plumed_coordinates: Optional[np.ndarray] = None  # whatever the array is!
```

File: mlptrain/configurations/configuration_set.py [hunk 2]
```diff
@@ -359,6 +403,35 @@ def _coordinates(self) -> np.ndarray:
...
         return np.array([np.asarray(c.coordinates, dtype=float) for c in self])
 
+    @property
+    def plumed_coordinates(self) -> Optional[np.ndarray]:
+        """
+        PLUMED collective variable values in this set
+
+        -----------------------------------------------------------------------
+        Returns:
+            (np.ndarray): PLUMED collective variable tensor (n, n_cvs)
```
Comment (t-young31) [inline mlptrain/configurations/configuration_set.py:413]: if the shape is `(n, n_cvs)` then it's a matrix
what's `n`?

File: mlptrain/configurations/configuration_set.py [hunk 3]
```diff
@@ -359,6 +403,35 @@ def _coordinates(self) -> np.ndarray:
...
+        """
+
+        n_cvs = 0
+        all_coordinates = []
+
+        for config in self:
+            all_coordinates.append(config.plumed_coordinates)
+
+            if config.plumed_coordinates is not None:
+                n_cvs = len(config.plumed_coordinates)
```
Comment (t-young31) [inline mlptrain/configurations/configuration_set.py:423]: I assume bad things happen if `n_cvs` is different for different `config`s ?

File: mlptrain/sampling/md.py [hunk 4]
```diff
@@ -282,37 +294,104 @@ def save_trajectory():
...
+    """Append unbiased energy (biased MLP energy - bias energy) to the
+    trajectory"""
 
-def _save_trajectory(ase_traj: 'ase.io.trajectory.Trajectory',
-                     traj_name: str,
-                     **kwargs
-                     ) -> None:
-    """Saves the trajectory with a unique name based on the current simulation
+    if isinstance(bias, PlumedBias):
+        energy = calculator.calc.get_potential_energy(ase_atoms)
```
Comment (t-young31) [inline mlptrain/sampling/md.py:371]: not a big fan of a `calc` attribute of a `calculator`. A calculator with or without a bias really should have the same interface i.e. a `get_potential_energy(...)` function

File: mlptrain/sampling/md.py [hunk 5]
```diff
@@ -282,37 +294,104 @@ def save_trajectory():
...
+def _append_biased_energy(ase_atoms, biased_energies, bias) -> None:
+    """Append bias energy to the trajectory"""
+
+    biased_energy = ase_atoms.get_potential_energy()
+
+    # Plumed calculator when get_potential_energy() is called returns an array
+    if isinstance(bias, PlumedBias):
+        biased_energy = biased_energy[0]
+
+    biased_energies.append(biased_energy)
```
Comment (t-young31) [inline mlptrain/sampling/md.py:389]: is this better? maybe

```suggestion
    result = ase_atoms.get_potential_energy()

    # Plumed calculator when get_potential_energy() is called returns an array
    biased_energies.append(result[0] if isinstance(bias, PlumedBias) else result)
```

Comment (t-young31) [reply mlptrain/sampling/md.py:389]: would be better if `get_potential_energy` returned something consistent! if this is an ASE problem I would consider subclassing the calculator and overriding `get_potential_energy`

File: mlptrain/sampling/md.py [hunk 6]
```diff
@@ -361,7 +440,7 @@ def _get_traj_name(restart_files: Optional[List[str]] = None,
                 return traj_name
 
 
-def _convert_ase_traj(traj_name: str) -> 'mlptrain.Trajectory':
+def _convert_ase_traj(traj_name, bias, **kwargs) -> 'mlptrain.Trajectory':
```
Comment (t-young31) [inline mlptrain/sampling/md.py:443]: still think it's worth type hinting all functions where possible

**Review by t-young31 — COMMENTED** [review]
awesome work – I got about halfway through and may or may not come back to it!

File: mlptrain/sampling/plumed.py [hunk 7]
```diff
@@ -28,63 +85,132 @@ def __init__(self,
...
-                raise TypeError('Supplied CVs are in incorrect format')
+            cvs = self._check_cvs_format(cvs)
+            self.cvs = cvs
 
         else:
             raise TypeError('PLUMED bias instantiation requires '
                             'a list of collective variables (CVs) '
                             'or a file containing PLUMED-type input')
 
+    # def __deepcopy__(self, memo=None):
```
Comment (tristan-j-wood) [inline mlptrain/sampling/plumed.py:114]: Remove lines

File: mlptrain/sampling/plumed.py [hunk 8]
```diff
@@ -150,8 +410,179 @@ def _from_file(self, file_name) -> None:
...
+
+        else:
+            raise TypeError('Supplied CVs are in incorrect format')
+
+        return cvs
+
+    def adjust_potential_energy(self, atoms) -> float:
+        """Adjust the energy of the system by adding PLUMED bias"""
+
+        # energy_bias = self.plumed_calculator.get_property(name='energy_bias',
```
Comment (tristan-j-wood) [inline mlptrain/sampling/plumed.py:570]: Remove

File: mlptrain/sampling/plumed.py [hunk 9]
```diff
@@ -150,8 +410,179 @@ def _from_file(self, file_name) -> None:
...
+        #                                                   atoms=atoms)
+        energy_bias = atoms.calc.get_property('energy_bias', atoms)
+        return energy_bias
+
+    def adjust_forces(self, atoms, forces) -> None:
+        """Adjust the forces of the system by adding PLUMED forces"""
+
+        # forces_bias = self.plumed_calculator.get_property(name='forces_bias',
```
Comment (tristan-j-wood) [inline mlptrain/sampling/plumed.py:578]: Remove

**Review by tristan-j-wood — APPROVED** [review]

**Pull request merged**